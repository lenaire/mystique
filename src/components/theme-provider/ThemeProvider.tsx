import { FC } from "react";
import {
  DefaultTheme, // This is overridden in types/theme.d.ts
  ThemeProvider as StyledComponentsThemeProvider,
  ThemeProviderProps as StyledComponentsThemeProviderProps,
} from "styled-components";

import { defaultTheme } from "$cmp/theme-provider/theme";
import { ResetGlobalCSS } from "$cmp/theme-provider/ResetGlobalCSS";

export type ThemeProviderProps = Partial<StyledComponentsThemeProviderProps<DefaultTheme>>;

export const ThemeProvider: FC<ThemeProviderProps> = ({ children, theme = defaultTheme }) => {
  return (
    <StyledComponentsThemeProvider theme={theme}>
      <ResetGlobalCSS />
      {children}
    </StyledComponentsThemeProvider>
  );
};
