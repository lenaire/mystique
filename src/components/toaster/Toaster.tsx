import { FC } from "react";
import { createPortal } from "react-dom";
import styled, {
  Keyframes,
  css,
  FlattenSimpleInterpolation,
  DefaultTheme,
} from "styled-components";
import { CmpProps } from "$types";
import { Icon, IconType } from "$cmp/icon/Icon";
import { Direction, Position, NotificationType } from "$utility/enums";
import { useToaster } from "./internals/useToaster";
import { slideInUp, slideInDown, slideInLeft, slideInRight } from "$utility/animations";

const TransitionMap: Record<string, Keyframes> = {
  [Direction.Up]: slideInUp,
  [Direction.Down]: slideInDown,
  [Direction.Left]: slideInLeft,
  [Direction.Right]: slideInRight,
};

const IconMap: Record<string, { color: (theme: DefaultTheme) => string; icon: IconType }> = {
  [NotificationType.Info]: {
    color: (theme: DefaultTheme): string => theme.palette.info.main,
    icon: "alert",
  },
  [NotificationType.Success]: {
    color: (theme: DefaultTheme): string => theme.palette.success.main,
    icon: "check",
  },
  [NotificationType.Error]: {
    color: (theme: DefaultTheme): string => theme.palette.error.main,
    icon: "error",
  },
  [NotificationType.Warning]: {
    color: (theme: DefaultTheme): string => theme.palette.warning.main,
    icon: "error",
  },
};

const PositionMap: Record<string, FlattenSimpleInterpolation> = {
  [Position.TopRight]: css`
    top: 1em;
    right: 1em;
  `,
  [Position.TopLeft]: css`
    top: 1em;
    left: 1em;
  `,
  [Position.BottomRight]: css`
    bottom: 1em;
    right: 1em;
  `,
  [Position.BottomLeft]: css`
    bottom: 1em;
    left: 1em;
  `,
};

const StyledIcon = styled(Icon)<{ $type: NotificationType }>`
  fill: ${(props): string => IconMap[props.$type].color(props.theme)};
  padding-right: 5px;
`;

const Toast = styled.div<{
  isActive: boolean;
  direction: Direction;
  duration: number;
  position: Position;
}>`
  position: fixed;
  top: 1em;
  right: 1em;
  padding: 4px;
  background: ${(props): string => props.theme.palette.common.white};
  width: 312px;
  height: 80px;
  border-radius: 8px;
  z-index: 10000;
  border: 1px solid ${(props): string => props.theme.palette.grey[300]};
  box-shadow: 0px 4px 4px rgba(177, 144, 198, 0.05), 0px 0px 20px rgba(177, 144, 198, 0.15);
  ${(props): Keyframes | false | FlattenSimpleInterpolation | string | undefined =>
    props.isActive
      ? css`
          animation: ${TransitionMap[props.direction]} 0.5s;
          animation-delay: 0s, ${props.duration}ms;
        `
      : css`
          opacity: 0;
          transition: visibility 0s, opacity 0.5s linear;
        `};
  ${(props): FlattenSimpleInterpolation => PositionMap[props.position]};
`;

const Header = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
`;

const Container = styled.div``;

const Body = styled.div`
  padding: 0 6px;
  display: flex;
  align-items: center;
`;

const getRootElement = (id: string): HTMLElement => document.getElementById(id) as HTMLElement;

export interface ToasterProps extends CmpProps {
  portalId: string;
  position?: Position;
  type?: NotificationType;
  direction?: Direction;
  onClose?: (event?: React.MouseEvent<HTMLButtonElement>) => void;
}

export const Toaster: FC<ToasterProps> = ({
  portalId,
  className,
  type = NotificationType.Info,
  direction = Direction.Right,
  position = Position.TopRight,
  onClose,
  ...rest
}) => {
  const { toasts, duration } = useToaster({});

  return createPortal(
    <Container className={className} {...rest}>
      {toasts.map((toast: any) => {
        return (
          <Toast
            key={toast.id}
            isActive={toast.active}
            duration={duration}
            direction={direction}
            position={position}
          >
            <Header>
              <button
                onClick={(event): void => {
                  onClose?.(event);
                }}
              >
                <Icon icon="close" />
              </button>
            </Header>
            <Body>
              <StyledIcon $type={type} icon={IconMap[type].icon} />
              {toast.content}
            </Body>
          </Toast>
        );
      })}
    </Container>,
    getRootElement(portalId)
  );
};
