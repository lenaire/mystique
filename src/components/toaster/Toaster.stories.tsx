import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "$cmp/button/Button";
import { useToaster } from "./internals/useToaster";
import { Toaster } from "./Toaster";

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: "Toaster",
  component: Toaster,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    className: { control: { disable: true } },
  },
} as ComponentMeta<typeof Toaster>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof Toaster> = (args) => {
  const { actions } = useToaster({ duration: 5000 });
  return (
    <div>
      <Button
        type="primary"
        label="Add Toast"
        onClick={(): void => actions.addToast("New Toast")}
      />
      <Toaster {...args} />
    </div>
  );
};

export const Primary = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Primary.args = {
  portalId: "toasts",
};
