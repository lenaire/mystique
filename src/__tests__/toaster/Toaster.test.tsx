import renderer from "react-test-renderer";
import { ThemeProvider } from "$cmp/theme-provider/ThemeProvider";
import { Toaster, ToasterProps } from "$cmp/toaster/Toaster";

jest.mock("react-dom", () => ({
  ...jest.requireActual("react-dom"),
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  createPortal: (node: React.ReactNode): React.ReactPortal => node,
}));

describe("<Icon />", () => {
  it("renders correctly", () => {
    const props: ToasterProps = {
      portalId: "toasts",
    };

    const tree = renderer
      .create(
        <ThemeProvider>
          <Toaster {...props} />
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
