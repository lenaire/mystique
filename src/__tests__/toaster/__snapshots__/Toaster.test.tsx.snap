// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toaster /> renders correctly 1`] = `
<div
  className="sc-pyfCe"
/>
`;
